File: web/src/api/identity.ts
import { request } from './http'

export type IdentityRole = 'node_admin' | 'tenant_admin' | 'operator' | 'viewer'
export type IdentityStatus = 'active' | 'disabled'

export type Tenant = {
  id: string
  name: string
  status: IdentityStatus
  created_at: string
  updated_at: string
}

export type OrgUnit = {
  id: string
  tenant_id: string
  parent_id?: string
  name: string
  created_at: string
  children?: OrgUnit[]
}

export type IdentityUser = {
  id: string
  tenant_id: string
  username: string
  display_name: string
  status: IdentityStatus
  all_orgs: boolean
  roles: string[]
  org_ids: string[]
  created_at: string
  updated_at: string
}

export type CreateUserInput = {
  tenant_id?: string
  username: string
  password: string
  display_name: string
  all_orgs?: boolean
  roles: string[]
  org_ids: string[]
}

export type UpdateUserInput = {
  display_name?: string
  status?: IdentityStatus
  all_orgs?: boolean
  roles?: string[]
  org_ids?: string[]
}

export const ROLE_META: Record<string, { label: string; hint: string }> = {
  node_admin: { label: '节点管理员', hint: '全部权限，含租户/用户/组织管理' },
  tenant_admin: { label: '租户管理员', hint: '设备全权、组织架构管理，Access 查看与确认' },
  operator: { label: '操作员', hint: '设备查看与启用，Access 查看' },
  viewer: { label: '观察者', hint: '设备与 Access 只读' },
}

// --- tenants ---

export function listTenants(): Promise<Tenant[]> {
  return request('/api/v1/tenants')
}

export function createTenant(name: string): Promise<Tenant> {
  return request('/api/v1/tenants', { method: 'POST', body: JSON.stringify({ name }) })
}

export function setTenantStatus(id: string, status: IdentityStatus): Promise<Tenant> {
  return request(`/api/v1/tenants/${id}`, { method: 'PATCH', body: JSON.stringify({ status }) })
}

// --- org units ---

export function listOrgUnits(tenantID?: string): Promise<OrgUnit[]> {
  const suffix = tenantID ? `?tenant_id=${encodeURIComponent(tenantID)}` : ''
  return request(`/api/v1/org-units${suffix}`)
}

export function createOrgUnit(tenantID: string | undefined, parentID: string, name: string): Promise<OrgUnit> {
  const suffix = tenantID ? `?tenant_id=${encodeURIComponent(tenantID)}` : ''
  return request(`/api/v1/org-units${suffix}`, {
    method: 'POST',
    body: JSON.stringify({ parent_id: parentID, name }),
  })
}

export function renameOrgUnit(id: string, name: string): Promise<OrgUnit> {
  return request(`/api/v1/org-units/${id}`, { method: 'PATCH', body: JSON.stringify({ name }) })
}

export function moveOrgUnit(id: string, parentID: string): Promise<OrgUnit> {
  return request(`/api/v1/org-units/${id}`, { method: 'PATCH', body: JSON.stringify({ parent_id: parentID }) })
}

export function deleteOrgUnit(id: string): Promise<void> {
  return request(`/api/v1/org-units/${id}`, { method: 'DELETE' })
}

// --- users ---

export function listUsers(tenantID?: string): Promise<IdentityUser[]> {
  const suffix = tenantID ? `?tenant_id=${encodeURIComponent(tenantID)}` : ''
  return request(`/api/v1/users${suffix}`)
}

export function createUser(input: CreateUserInput): Promise<IdentityUser> {
  return request('/api/v1/users', { method: 'POST', body: JSON.stringify(input) })
}

export function updateUser(id: string, input: UpdateUserInput): Promise<IdentityUser> {
  return request(`/api/v1/users/${id}`, { method: 'PATCH', body: JSON.stringify(input) })
}

export function deleteUser(id: string): Promise<void> {
  return request(`/api/v1/users/${id}`, { method: 'DELETE' })
}

export function setUserPassword(id: string, password: string): Promise<void> {
  return request(`/api/v1/users/${id}/password`, {
    method: 'POST',
    body: JSON.stringify({ password }),
  })
}

// --- roles ---

export function listRoles(): Promise<{ roles: string[] }> {
  return request('/api/v1/roles')
}

File: internal/identity/identity.go
package identity

import (
	"context"
	"time"

	"github.com/jackc/pgx/v5/pgxpool"
)

type Tenant struct {
	ID        string    `json:"id"`
	Name      string    `json:"name"`
	Status    string    `json:"status"`
	CreatedAt time.Time `json:"created_at"`
	UpdatedAt time.Time `json:"updated_at"`
}

// OrgUnit is a node in the per-tenant organization tree. Top-level nodes
// have a nil ParentID; the tree starts empty (no seed root) and is built
// by users. It is the data-scope anchor: users get scopes pointing at org
// units (subtree expansion), devices optionally point at an org unit.
type OrgUnit struct {
	ID        string     `json:"id"`
	TenantID  string     `json:"tenant_id"`
	ParentID  *string    `json:"parent_id,omitempty"`
	Name      string     `json:"name"`
	CreatedAt time.Time  `json:"created_at"`
	Children  []*OrgUnit `json:"children,omitempty"`
}

type User struct {
	ID           string    `json:"id"`
	TenantID     string    `json:"tenant_id"`
	Username     string    `json:"username"`
	DisplayName  string    `json:"display_name"`
	PasswordHash string    `json:"-"`
	Status       string    `json:"status"`
	AllOrgs      bool      `json:"all_orgs"`
	Roles        []string  `json:"roles"`
	OrgIDs       []string  `json:"org_ids"`
	CreatedAt    time.Time `json:"created_at"`
	UpdatedAt    time.Time `json:"updated_at"`
}

type CreateTenantInput struct {
	Name string `json:"name"`
}

func (in CreateTenantInput) Validate() error {
	if in.Name == "" || len(in.Name) > 255 {
		return invalid("tenant name must be non-empty and at most 255 bytes")
	}
	return nil
}

// CreateUserInput creates a user in the caller's tenant by default. A
// node_admin may set TenantID to create a user in another tenant (enforced
// by the handler). Roles, OrgIDs and AllOrgs may be empty: a user can be
// created first and have roles/org scopes assigned later via Update.
type CreateUserInput struct {
	TenantID    string   `json:"tenant_id"`
	Username    string   `json:"username"`
	Password    string   `json:"password"`
	DisplayName string   `json:"display_name"`
	AllOrgs     bool     `json:"all_orgs"`
	Roles       []string `json:"roles"`
	OrgIDs      []string `json:"org_ids"`
}

func (in CreateUserInput) Validate() error {
	if in.Username == "" || len(in.Username) > 64 {
		return invalid("username must be non-empty and at most 64 bytes")
	}
	if in.Password == "" || len(in.Password) > 256 {
		return invalid("password must be between 1 and 256 bytes")
	}
	if len(in.DisplayName) > 255 {
		return invalid("display_name must be at most 255 bytes")
	}
	for _, role := range in.Roles {
		if !ValidRole(role) {
			return invalid("unknown role: " + role)
		}
	}
	return nil
}

type UpdateUserInput struct {
	DisplayName *string  `json:"display_name"`
	Status      *string  `json:"status"`
	AllOrgs     *bool    `json:"all_orgs"`
	Roles       []string `json:"roles"`
	OrgIDs      []string `json:"org_ids"`
}

const (
	RoleNodeAdmin      = "node_admin"
	RoleTenantAdmin    = "tenant_admin"
	RoleOperator       = "operator"
	RoleViewer         = "viewer"
	UserStatusActive   = "active"
	UserStatusDisabled = "disabled"
)

func ValidRole(role string) bool {
	switch role {
	case RoleNodeAdmin, RoleTenantAdmin, RoleOperator, RoleViewer:
		return true
	}
	return false
}

// AllRoles returns the fixed role names in a stable order.
func AllRoles() []string {
	return []string{RoleNodeAdmin, RoleTenantAdmin, RoleOperator, RoleViewer}
}

// PrincipalInfo is the minimal caller identity that management handlers
// need. It is populated by the nodeapp wiring from authn.Principal, keeping
// this package free of an authn dependency (authn imports identity).
type PrincipalInfo struct {
	UserID   string
	TenantID string
}

type TenantRepository interface {
	Create(context.Context, CreateTenantInput) (Tenant, error)
	List(context.Context) ([]Tenant, error)
	Get(context.Context, string) (Tenant, error)
	GetByName(context.Context, string) (Tenant, error)
	SetStatus(context.Context, string, string) (Tenant, error)
}

type OrgUnitRepository interface {
	Create(context.Context, string, *string, string) (OrgUnit, error)
	Tree(context.Context, string) ([]*OrgUnit, error)
	Get(context.Context, string, string) (OrgUnit, error)
	UpdateName(context.Context, string, string, string) (OrgUnit, error)
	UpdateParent(context.Context, string, string, *string) (OrgUnit, error)
	UpdateNameAndParent(context.Context, string, string, string, *string) (OrgUnit, error)
	Delete(context.Context, string, string) error
	SubtreeIDs(context.Context, []string) ([]string, error)
}

type UserRepository interface {
	Create(context.Context, CreateUserInput, string) (User, error)
	List(context.Context, string) ([]User, error)
	Get(context.Context, string, string) (User, error)
	GetByUsername(context.Context, string, string) (User, error)
	Update(context.Context, string, string, UpdateUserInput) (User, error)
	Delete(context.Context, string, string) error
	SetPassword(context.Context, string, string, string) error
}

type Store struct {
	Tenants  TenantRepository
	OrgUnits OrgUnitRepository
	Users    UserRepository
}

func NewStore(pool *pgxpool.Pool) *Store {
	return &Store{
		Tenants:  NewPostgresTenantRepository(pool),
		OrgUnits: NewPostgresOrgUnitRepository(pool),
		Users:    NewPostgresUserRepository(pool),
	}
}
